Separate out validation of groups of settings

cbismuth opened the pull request:
This pull request allows a setting to be validated against target settings to ensure runtime dependencies are met (e.g. disk watermarks `low`, `high` and `flood_stage`).

This pull request also includes a fix to not update settings when no value has changed.

elasticmachine commented:
Pinging @elastic/es-core-infra

DaveCTurner commented:
@elasticmachine test this please

cbismuth commented:
Sure, I'll report the test you've suggested and add a new one on this change in particular, thank you @DaveCTurner.

DaveCTurner commented:
Note that [the build that failed](
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request+oss-distro-docs/1359/) looks like it failed for good reason. Could you move this failure into a unit test and then fix it?

cbismuth commented:
Yes :+1:

cbismuth commented:
> Note that [the build that failed](https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request+oss-distro-docs/1359/) looks like it failed for good reason. Could you move this failure into a unit test and then fix it?

Issue found and fixed in f77d1e7, thanks @DaveCTurner. I'm now focusing on implementing relevant test cases.

DaveCTurner commented:
Responding to https://github.com/elastic/elasticsearch/issues/28309#issuecomment-42700826 here since this is the right place to discuss this PR.

The single test present in this PR (the current head that I'm looking at is f77d1e70089f5af03e7f47c54d6ba2b75959dc51) does not support the change to the production code, in the sense that it passes on today's `master` as well. To be worthwhile it needs to be a test that fails without the change in place.

The test failure that you found and fixed should also be captured in a test.

Additionally I [think we should have a test that focusses on this change to the way that Settings work](
https://github.com/elastic/elasticsearch/pull/34184#pullrequestreview-161546206) which is independent of the disk watermark feature.

cbismuth commented:
I agree with you @DaveCTurner, tests are under development.

Let me please reword: in my previous comment, I meant I've modified [the test you suggested](https://github.com/elastic/elasticsearch/pull/34184/files#diff-6a01a97f6f49e773fe8a6396575474ecR217) and I wanted to know whether or not you agree with the changes I've made in your test. Do you see what I mean?

DaveCTurner commented:
> I wanted to know whether or not you agree with the changes I've made in your test. Do you see what I mean?

I think so, and my feedback was that this test passes on `master` so doesn't demonstrate the bug that we're chasing, which means it's not the test we need for this change. I suppose it's a reasonable test in the sense that it's expected to pass, but I also think there are already test cases that state the same thing more generally so it's not really necessary.

If there is more work in progress then it's probably best to ping when the whole PR is ready to review - it'll be easier to see the whole picture.

cbismuth commented:
I understand, thank you.

cbismuth commented:
@DaveCTurner PR is complete and ready for review.

cbismuth commented:
I've found another bug in the `AbstractScopedSettings#updateSettings` API: the return value was `true` even when no value has changed.

I've added commit 68044ba to fix it and I've added a test case to cover this change.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettingsTests.java:
```diff
@@ -203,4 +203,20 @@ public void testInvalidHighDiskThreshold() {
+        final ClusterSettings clusterSettings = new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        new DiskThresholdSettings(settings, clusterSettings); // this has the effect of registering the settings updater
+
+        settings = clusterSettings.applySettings(Settings.builder()
```
I see what's happened here, and I must apologise. The unit test I suggested was looking at `applySettings` but really I should have used `updateSettings` since this is where the update is calculated. My bad.

However, this test still passes on `master` without any changes so I don't think it should be added here. Could you make an appropriate test based around `updateSettings` instead, and check that it fails on `master` and passes on this branch?

cbismuth replied:
No worries at all. Sure, I'll update the test and run it on this branch and on `master`.

DaveCTurner commented on server/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java:
```diff
@@ -1221,4 +1256,54 @@ public String getKey(final String key) {
+            updates,
+            "transient"
+        );
+        assertFalse(updated);
```
I think it's ok to return `true` here. Claiming to have updated a setting when really it hasn't changed in value is a right-side failure: it means we call `reroute()` to reallocate any shards affected by the new settings, but this does nothing if the setting change was a no-op. On the other hand failing to call `reroute()` when a setting _does_ change is very bad, so changing the behaviour here needs a lot more scrutiny.

cbismuth replied:
I understand. I'll restore the previous behavior to not break the current data flow and leave this responsibility to downstream components, it's safer.

I wasn't very at ease with this change, thank you for catching it. I'll dig deeper into it with a debugger just for my own curiosity.

DaveCTurner commented on server/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java:
```diff
@@ -1221,4 +1256,54 @@ public String getKey(final String key) {
+        );
+        assertTrue(updated);
+        assertThat(target.get(SETTING_FOO_LOW.getKey()), equalTo("20"));
+        assertThat(updates.get(SETTING_FOO_LOW.getKey()), equalTo("20"));
```
We can write this with fewer lines and fewer mutable variables by putting each of these steps into its own block like this:

```
        {
            final Settings.Builder updates = Settings.builder();
            assertTrue(service.updateSettings(Settings.builder().put(SETTING_FOO_LOW.getKey(), 20).build(), target, updates, "transient"));
            assertThat(target.get(SETTING_FOO_LOW.getKey()), equalTo("20"));
            assertThat(updates.get(SETTING_FOO_LOW.getKey()), equalTo("20"));
        }
```

The line-width limit is 140 characters, and this fits into that very neatly. It's something of a matter of taste, but we seem to prefer denser code in this sort of situation.

cbismuth replied:
That's fine for me. The more immutable code is, the more confident I feel.

DaveCTurner commented on server/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java:
```diff
@@ -1221,4 +1256,54 @@ public String getKey(final String key) {
+            Settings.builder(),
+            "transient"
+        ));
+        assertThat(exception.getMessage(), equalTo("[high]=10 is lower than [low]=20"));
```
Could we also have a case that shows that if `low` is set below `10` then `high` can also be set below `10`, to verify that we're not validating against both the default and the actual values?

DaveCTurner commented on server/src/test/java/org/elasticsearch/common/settings/ScopedSettingsTests.java:
```diff
@@ -54,11 +54,46 @@
 
 public class ScopedSettingsTests extends ESTestCase {
 
+    private static class FooLowSettingValidator implements Setting.Validator<Integer> {
```
I think these settings can reasonably be moved right next to `testUpdateOfValidationDependantSettings` since that's the only place they're used.

cbismuth replied:
Yes, that would be better, I'll move them.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java:
```diff
@@ -716,10 +719,12 @@ private boolean updateSettings(Settings toApply, Settings.Builder target, Settin
-                changed = true;
+                changed = hasChanged(toApply, target);
+                if (changed) {
+                    validate(toApply, target.build());
```
I think this isn't quite the right approach. As I understand it this change means we're validating the whole settings update here, but by my reading we already validate that later on in the process:

https://github.com/elastic/elasticsearch/blob/4dc3adad5172f1d7c137b983a1715797a9021190/server/src/main/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdater.java#L94-L95

I think it'd be better to try and do _less_ validation here on settings that have dependent settings. Perhaps the `Setting.Validator` interface should have two `validate()` methods, one for this stage of validation (good for settings with no dependencies, i.e., most of them) and one for the full validation including dependencies.

cbismuth replied:
Yes, I see what you mean. I'll try to move the full validation including dependencies in a new API in `Setting.Validator`. I also think it's a better location for this.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java:
```diff
@@ -733,6 +738,35 @@ private boolean updateSettings(Settings toApply, Settings.Builder target, Settin
+
+    private boolean hasChanged(Settings toApply, Settings.Builder target) {
+        boolean changed = toApply.keySet().stream().anyMatch(k -> k.endsWith("*"));
+        if (!changed) {
```
The house style is to avoid using the `!` unary operator and to prefer saying `changed == false` here, because `!` is small and easy to miss in a long expression, and `== false` is larger and therefore more visible. I think this code will go away due to other comments, but thought I'd raise it anyway for your info.

cbismuth commented:
Thank you @DaveCTurner for the kind words and insightful review. I'll rework my changes and let you know when it's ready.

cbismuth commented:
Hi @DaveCTurner, I've updated this PR with all your suggestions.

I've added a lightweight validation API to the `Setting.Validator` interface, and I also think the code looks better now (validation scope and readability).

I've also found an existing `ScopedSettingsTests#testValidator` that I didn't see before. It is now obsolete, I've dropped it as the new one has a greater coverage and has a more readable data flow.

The test you suggested has been updated, it passes on this branch and it fails on `master` with the error below, that's good news.

```
java.lang.IllegalArgumentException: high disk watermark [97%] more than flood stage disk watermark [95%]
```

I've reviewed every change I've made and rerun all tests, PR is ready for review.

cbismuth commented:
I'm working on fixing the build.

cbismuth commented:
Hi, this PR is synced with `master` and tests are green again.

Could you please run a CI build? I stay available for any further change, thank you.

andyb-elastic commented:
Jenkins, ok to test

It should run the build when you push a new commit now

cbismuth commented:
Nice, thank you @andyb-elastic :+1:

cbismuth commented:
Tests are green, PR is ready for review, thank you.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/settings/Setting.java:
```diff
@@ -1037,7 +1054,7 @@ public void apply(T value, Settings current, Settings previous) {
     }
 
-    public static Setting<String> simpleString(String key, Validator<String> validator, Property... properties) {
+    public static Setting<String> simpleStringWithValidator(String key, Validator<String> validator, Property... properties) {
```
I think we could leave this called `simpleString`. The convention seems to be not to call out the existence of the Validator explicitly in other calls. I see that this might cause some issues with overloading but I think they're surmountable.

cbismuth replied:
Exactly, I'll try to work around overloading issues.

cbismuth replied:
The build is setup to disallow any compiler warning.

I've tried to work around this issue by having `simpleString` API with different number of functional interface parameters, but the warning is still there and I don't understand why. I've posted a [question on SO](https://stackoverflow.com/questions/53080812/java-functional-interfaces-weird-potentially-ambiguous-warning) to get some feedback on this issue, I'll let you know.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java:
```diff
@@ -72,14 +72,14 @@
     public static final Setting.AffixSetting<String> CLUSTER_ROUTING_REQUIRE_GROUP_SETTING =
-        Setting.prefixKeySetting(CLUSTER_ROUTING_REQUIRE_GROUP_PREFIX + ".", (key) ->
-            Setting.simpleString(key, (value, map) -> IP_VALIDATOR.accept(key, value), Property.Dynamic, Property.NodeScope));
+        Setting.prefixKeySetting(CLUSTER_ROUTING_REQUIRE_GROUP_PREFIX + ".", (key) -> Setting.simpleStringWithValidator(
```
I think we can restore the whitespace to approximately what it was beforehand.

cbismuth replied:
Sure.

cbismuth replied:
The `simpleStringWithValidator` new API name make these lines longer than 140 characters which makes Checkstyle validation fail.

I'll restore the whitespace as soon as I'll be able restore the preferred `simpleString` API name.

cbismuth replied:
Temporary fixed in 43840f138129d15ec4322f03db33379631cb3a4b.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/settings/Setting.java:
```diff
@@ -813,13 +822,21 @@ public String getNamespace(Setting<T> concreteSetting) {
     public interface Validator<T> {
 
+        /**
+         * The validation routine for this validator.
```
```suggestion
         * Validate this setting's value in isolation.
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/settings/Setting.java:
```diff
@@ -813,13 +822,21 @@ public String getNamespace(Setting<T> concreteSetting) {
+        void validate(T value);
+
         /**
          * The validation routine for this validator.
```
```suggestion
         * Validate this setting against its dependencies, specified by {@link #settings()}.
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdaterTests.java:
```diff
@@ -471,4 +449,108 @@ public void testRemovingArchivedSettingsDoesNotRemoveNonArchivedInvalidOrUnknown
+        return invalidSettings;
+    }
+
+    private static Setting<String> createInvalidSetting(int index) {
```
Could we distinguish invalid-in-isolation settings from invalid-with-dependencies settings to make sure we're doing both kinds of validation as appropriate?

cbismuth replied:
Yes, it will be better :+1:

cbismuth replied:
Fixed in https://github.com/elastic/elasticsearch/pull/34184/commits/2efdf1e79f66e43f5b9a1450542b85f018d6a6ae#diff-7cc27247465d367dd79b97b65722eae1R463 with random creation pattern.

DaveCTurner commented on server/src/test/java/org/elasticsearch/common/settings/SettingTests.java:
```diff
@@ -206,6 +206,10 @@ public void testValidateStringSetting() {
         public static boolean invoked;
 
+        @Override
+        public void validate(String value) {
```
I think this method should also contain assertions and we should also assert that it is called in the right places.

cbismuth replied:
Yes, you're right, I'll have a look at it.

cbismuth replied:
Good catch! Fixed in 3e58473, thank you :+1:

DaveCTurner commented on server/src/test/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdaterTests.java:
```diff
@@ -471,4 +449,108 @@ public void testRemovingArchivedSettingsDoesNotRemoveNonArchivedInvalidOrUnknown
+    private static final Setting<Integer> SETTING_FOO_HIGH = new Setting<>("foo.high", "100",
+        Integer::valueOf, new FooHighSettingValidator(), Property.Dynamic, Setting.Property.NodeScope);
+
+    public void testUpdateOfValidationDependantSettings() {
```
```suggestion
    public void testUpdateOfValidationDependentSettings() {
```

Just spotted a typo in the test name.

DaveCTurner commented:
> Maybe we should update this PR title with something related to your JavaDoc fixes: "Validate settings in isolation and against their dependencies", don't you think?

Yes, it's useful if the PR title and body reflect the title and body of the commit message that we end up using on merge. That's normally something I'd take care of at merge time if needed, but here's what I'd change them to at the moment (within the constraints of the 50/72 rule).

```
Separate out validation of groups of settings

We ensure that every settings update results in a valid state using a
`Setting.Validator`. Most implementations of this check that each
setting is valid in isolation, but there are a few groups of settings
each of which can affect the validity of the others and this can also be
checked using a `Setting.Validator`.

We validate the settings updates in two phases: firstly we check each
setting in turn, ignoring dependent settings, and then we validate the
groups of dependent settings later. However, today, there is only one
`validate()` method which is used for both kinds of validation, and this
means we sometimes validate settings against the default values of their
dependencies, which may lead to an update being erroneously rejected in
the first phase.

This change splits the `validate()` method into two, so that the two
phases of validation can be distinguished.

Fixes #28309
```

cbismuth commented:
These are great PR title and body, thank you @DaveCTurner. I'll update the current ones and take these guidelines into account for my upcoming contributions.

DaveCTurner commented:
> I still can't rename simpleStringWithValidator to simpleString as I get a weird javac warning (see below) which make the build fail. Do you have an idea to work around this?

I guess `javac` has a point, it could be ambiguous. But the thing with which it's ambiguous is suspicious, it has an argument `Function<String, String> parser` which shouldn't be doing anything. As far as I can tell this is only used in one place, `RemoteClusterAware`, and is used to _validate_ that `parsePort` doesn't throw an exception, so maybe we can remove that overload and use a validator instead?

DaveCTurner commented on plugins/examples/custom-settings/src/main/java/org/elasticsearch/example/customsettings/ExampleCustomSettingsConfig.java:
```diff
@@ -49,7 +49,7 @@
      * A string setting that can be dynamically updated and that is validated by some logic
      */
-    static final Setting<String> VALIDATED_SETTING = Setting.simpleString("custom.validated", (value, settings) -> {
+    static final Setting<String> VALIDATED_SETTING = Setting.simpleString("custom.validated", (value) -> {
```
One nit: in all the places we now have a single-argument lambda, there's no need for the parentheses around the argument.

cbismuth commented:
Thanks @DaveCTurner, I’ll update this PR next week, I’m on holidays too ;-)

s1monw commented on server/src/main/java/org/elasticsearch/common/settings/Setting.java:
```diff
@@ -814,12 +823,20 @@ public String getNamespace(Setting<T> concreteSetting) {
          * @param settings a map from the settings specified by {@link #settings()}} to their values
          */
-        void validate(T value, Map<Setting<T>, T> settings);
+        default void validate(T value, Map<Setting<T>, T> settings) {
```
does it make sense to delegate to `validate(T value)` here?

s1monw replied:
from an interface perspective I think we might be able to only have `validate(T value, Optional<Map<Setting<T>, T>> settings)` and instead check the optional if needed?

cbismuth replied:
Thanks @s1monw.

I'd prefer to keep these two API as most clients will only be interested in validating a setting in isolation (i.e. not against dependencies).
Besides, we do a fail-fast validation [here](https://github.com/elastic/elasticsearch/pull/34184/files/e83e210c8d35d427b0f35d851e65b1169f6d2b43#diff-3b117ec54ae350c750ab41abc7e4327fR721) where we don't have resolved setting dependencies yet. Therefore, if we want to keep this fail-fast shortcut, any client will have to implement an `if optional present/absent` block even if he's interested in validating a setting against its dependencies.

What do you think about these two points?

DaveCTurner replied:
I think I prefer the two methods to the optional parameter too. In practice there are very few dependencies between settings, so the one-argument lambda is less noisy in almost all cases. Settings _with_ dependencies can't use a lambda anyway because they have to override the `settings()` function too.

I also think there's no need for the dependent validation to delegate to the other one, because we call both during settings validation. However I think this should be mentioned in the Javadocs as implementors shouldn't need to go to the implementation to check this (and I know I'd forget and have to check). Also the interface-level comment needs updating as it's no longer wholly accurate.

cbismuth replied:
Hi @DaveCTurner, here is a additional commit 7c33bfe6eb9cb43412bbae0a723657fd96e0a7b9 to improve the `Validator` class documentation, thank you.

cbismuth replied:
I meant _developers who will implement this interface_.

We tried to make javadoc more explicit, but I see what you mean: the API doesn't enforce the developer to ask himself whether or not he would like to implement dependencies validation. But as it's an edge case, I still hope the developer would read the javadoc or jump to the interface definition.

cbismuth commented:
Hi @s1monw and @DaveCTurner, I'm sure everyone is quite busy, so here is a quick follow up below to ease PR validation and merge.

* This PR separates out validation of groups of settings,
* A fail-fast validation is done early in the code to check each updated setting without resolving its dependencies,
* A full validation is done later to check dependencies of each updated setting,
* We've introduced a new API to validate dependencies of a setting and it is a no-op default method in the `Validator` class,
* @s1monw would prefer to have a single API to implement in the `Validator` interface without the no-op default interface method for setting dependencies validation.

We have three options:

1. Leave things as they are now and rely on the updated Java documentation to tell developers there's a no-op method to override to validate dependencies of a setting.
1. Refactor the `Validator` interface to have a single `validate(T value, Optional<Map<Setting<T>, T>> settings)` API and delegate to developers the task to check whether or not the optional parameter is present or absent (it will always be absent when the early fail-fast first check is done).
1. Refactor the `Validator` interface to have a single `validate(T value, Map<Setting<T>, T> settings)` API and remove the early first check so that the map parameter will always be present (not empty when a setting has dependencies or empty when a setting has no dependency).

I would prefer option 2, and you?

cbismuth commented:
Sorry, I made a mistake in my previous comment, the option I chose was zero-based ... so I would prefer option **3**, and you?

cbismuth commented:
We have a green build :+1:

cbismuth commented:
Hi, I know it's Wednesday and there're probably many things to close before the weekend.

I hope we will be able state whether or not this PR can be merged as it is (with latest [documentation fix](https://github.com/elastic/elasticsearch/pull/34184/commits/7c33bfe6eb9cb43412bbae0a723657fd96e0a7b9) included) or should we go for a new iteration to simplify the `Validator<T>` interface.

The more I look at it, the more I think this version is a good one and we should merge it as it is.

cbismuth commented:
Hi @DaveCTurner, what do you think we should do on this PR? Thanks.

DaveCTurner commented:
Hi @cbismuth, just to let you know that this PR is still in my queue. Bear with me, I have a few too many things in flight right now.

cbismuth commented:
No worries @DaveCTurner, I'll be available when needed.

cbismuth commented:
That's great! Thanks a lot @DaveCTurner for finely reviewing the code and no worries at all for the time it took.
